lib/datalog/roletester/src/role_tester.rs

```diff
@@ -0,0 +1,162 @@
+
+    DenyLogins(user, login, role) <- HasDeniedLogin(user, login, role);
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+struct Predicate {
+    atoms: Vec<u32>
+}
+
+#[derive(Serialize, Deserialize)]
+struct EDB {
+    has_role: Option<Vec<Predicate>>,
+	has_trait: Option<Vec<Predicate>>,
+	has_login_trait: Option<Vec<Predicate>>,
+	role_allows_login: Option<Vec<Predicate>>,
+	role_denies_login: Option<Vec<Predicate>>,
+	role_allows_node_label: Option<Vec<Predicate>>,
+	role_denies_node_label: Option<Vec<Predicate>>,
+	node_has_label: Option<Vec<Predicate>>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct IDB {
+    accesses: Vec<Predicate>,
+    deny_accesses: Vec<Predicate>,
+    deny_logins: Vec<Predicate>
+}
+
+#[no_mangle]
+pub extern "C" fn process_access(s: *const c_char) -> *mut c_char {
+    let mut runtime = Crepe::new();
+    let c_str = unsafe {
+        assert!(!s.is_null());
+        CStr::from_ptr(s)
+    };
+    let r_str = c_str.to_str().unwrap();
+    let edb: EDB = serde_json::from_str(r_str).unwrap();
+    
+    match edb.has_role {
+        Some(x) => {
+            x.into_iter()
+             .for_each(|Predicate{atoms}| runtime.extend(&[HasRole(atoms[0], atoms[1])]));
+        },
+        None => {}
+    }
+
+    match edb.has_trait {
+        Some(x) => {
+            x.into_iter()
+             .for_each(|Predicate{atoms}| runtime.extend(&[HasTrait(atoms[0], atoms[1], atoms[2])]));
+        },
+        None => {}
+    }
+    
+    match edb.role_allows_login {
+        Some(x) => {
+            x.into_iter()
+             .for_each(|Predicate{atoms}| runtime.extend(&[RoleAllowsLogin(atoms[0], atoms[1])]));
+        },
+        None => {}
+    }
+
+    match edb.role_denies_login {
+        Some(x) => {
+            x.into_iter()
+             .for_each(|Predicate{atoms}| runtime.extend(&[RoleDeniesLogin(atoms[0], atoms[1])]));
+        },
+        None => {}
+    }
+
+    match edb.role_allows_node_label {
+        Some(x) => {
+            x.into_iter()
+             .for_each(|Predicate{atoms}| runtime.extend(&[RoleAllowsNodeLabel(atoms[0], atoms[1], atoms[2])]));
+        },
+        None => {}
+    }
+
+    match edb.role_denies_node_label {
+        Some(x) => {
+            x.into_iter()
+             .for_each(|Predicate{atoms}| runtime.extend(&[RoleDeniesNodeLabel(atoms[0], atoms[1], atoms[2])]));
+        },
+        None => {}
+    }
+
+    match edb.node_has_label {
+        Some(x) => {
+            x.into_iter()
+             .for_each(|Predicate{atoms}| runtime.extend(&[NodeHasLabel(atoms[0], atoms[1], atoms[2])]));
+        },
+        None => {}
+    }
+
+    // Format output into JSON
+    let (accesses, deny_accesses, deny_logins) = runtime.run();
+    let idb = IDB {
+        accesses: accesses.into_iter().map(|HasAccess(a, b, c, d)| Predicate{atoms: vec![a, b, c, d]}).collect(),
```

Comment:
Bit of a nit but usually you use a `Predicate::new` associated function to perform repeated construction in many places like this.